Mathlib/Tactic/Recover.lean

```diff
@@ -0,0 +1,72 @@
+/-
+Copyright (c) 2022 Siddhartha Gadgil. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Gabriel Ebner
+-/
+import Lean
+import Mathlib.Tactic.Cache
+import Mathlib.Tactic.RCases
+import Std
+
+open Std (HashSet PHashSet)
+open Lean Meta Elab Tactic
+
+/- The code below (up to the tactic) is from Aesop due to Janis Limperg-/
+namespace Mathlib.Tactic
+
+namespace Std.HashSet
+
+/-- insert many elements in a HashSet (from Aesop) -/
+def insertMany [ForIn Id ρ α] [BEq α] [Hashable α] (s : HashSet α) (as : ρ) :
+    HashSet α := Id.run do
+  let mut s := s
+  for a in as do
+    s := s.insert a
+  return s
+
+end Std.HashSet
+
+/-- instantiating MVars in goal (from Aesop) -/
+def instantiateMVarsInGoal (mvarId : MVarId) : MetaM Unit := do
+  discard $ getMVarDecl mvarId
+    -- The line above throws an error if the `mvarId` is not declared. The line
+    -- below panics.
+  instantiateMVarDeclMVars mvarId
+
+/-- Returns unassigned metavariable dependencies (from Aesop) -/
+partial def getUnassignedGoalMVarDependencies (mvarId : MVarId) :
+    MetaM (HashSet MVarId) :=
+  return (← go mvarId |>.run {}).snd
+  where
+    /-- helper for adding metavariables -/
+    addMVars (e : Expr) : StateRefT (HashSet MVarId) MetaM Unit := do
+      let mvars ← getMVarsNoDelayed e
```

Comment:
Won't this just ignore metavariables in delayed assignments?  cc @jlimperg

Reply:
Yes, this looks like a bug. I'll update it in Aesop then link the new code.

Reply:
New code: https://github.com/JLimperg/aesop/pull/25

@gebner Could you give this another look?

Reply:
I've finally fixed this function.